File: Auxiliary_functions/PRS_CFO_estimate.m
function [rxWaveformFreqCorrected, varargout] = PRS_CFO_estimate(rxWaveform,refWaveform,TX_PRS_config,searchBW,varargin)
% Performs a search for central frequency offset (CFO) and PRS ID
%   (optionally).
% INPUTS:
%       rxWaveform - received waveform, vector N_RXsym x 1
%       refWaveform - reference waveform, vector N_RefSym x 1 or matrix N_RefSym x N_PRSid
%       TX_PRS_config - transmit/reference signal configuration file 
%       searchBW - bandwidth for the frequency offset search
%       prsIDlist - list for potential PRS IDs (if search is performed), or
%               known PRS ID (if search is not performed)
%   Optional input:
%       do_plot - indicator of plotting capability (true or false, false being default)
% OUTPUTS: 
%       rxWaveformFreqCorrected - frequency offset-corrected received
%           waveform
%       freqOffset - estimated frequency offset
%   Optional output:
%       prsID - found PRS ID (if search is performed)
%
if size(refWaveform,2) == 1
    doPRSSearch = false;
else
    doPRSSearch = true;
end
% Vector of potential PRS IDs to check against OR a known PRS ID
prsIDlist = varargin{1};
if numel(varargin) > 1
    do_plot = varargin{2};
else
    do_plot = false; 
end
%
% Set the CFO searching step
fshifts = (-searchBW:TX_PRS_config.subcarrierSpacing:searchBW) * 1e3 / 2; % Half subcarrier step
t = (0:size(rxWaveform,1)-1).' / TX_PRS_config.SampleRate;
% Calculate the correlation function between received and reference
% waveform, perform PRS ID (optional) and CFO search
if doPRSSearch
    peak_value = zeros(numel(fshifts),size(refWaveform,2));
    peak_index = zeros(numel(fshifts),size(refWaveform,2));
    for fIdx = 1:numel(fshifts)
        coarseFrequencyOffset = fshifts(fIdx);
        rxWaveformFreqCorrected = rxWaveform .* exp(-1i*2*pi*coarseFrequencyOffset*t);
        %         % Downsample to the minumum sampling rate to cover SSB bandwidth
        %         rxWaveformDS = resample(rxWaveformFreqCorrected,syncSR,rxSampleRate);
        T = numel(rxWaveform);
        for prsIdx = 1:size(refWaveform,2)
            refW = refWaveform(:,prsIdx);
            minlength = size(refWaveform,1);
            % Get the number of time samples T and receive antennas R in the
            % waveform
            if (T < minlength)
                waveformPad = [rxWaveformFreqCorrected; zeros(minlength-T,1)];
                T = minlength;
            else
                waveformPad = rxWaveformFreqCorrected;
            end
            refcorr = xcorr(waveformPad,refW);
            corr = abs(refcorr(T:end));
            
            [peak_value(fIdx,prsIdx),peak_index(fIdx,prsIdx)] = max(corr);
            peak_index(fIdx,prsIdx) = peak_index(fIdx,prsIdx) + TX_PRS_config.SymbolLengths(1);
            
        end
    end
else
    peak_value = zeros(numel(fshifts),1);
    peak_index = zeros(numel(fshifts),1);
    for fIdx = 1:numel(fshifts)
        coarseFrequencyOffset = fshifts(fIdx);
        rxWaveformFreqCorrected = rxWaveform .* exp(-1i*2*pi*coarseFrequencyOffset*t);
        %         % Downsample to the minumum sampling rate to cover SSB bandwidth
        %         rxWaveformDS = resample(rxWaveformFreqCorrected,syncSR,rxSampleRate);
        T = numel(rxWaveform);
        minlength = size(refWaveform,1);
        % Get the number of time samples T and receive antennas R in the
        % waveform
        if (T < minlength)
            waveformPad = [rxWaveformFreqCorrected; zeros(minlength-T,1)];
            T = minlength;
        else
            waveformPad = rxWaveformFreqCorrected;
        end
        refcorr = xcorr(waveformPad,refWaveform);
        corr = abs(refcorr(T:end));
        [peak_value(fIdx),peak_index(fIdx)] = max(corr);
        peak_index(fIdx) = peak_index(fIdx) + TX_PRS_config.SymbolLengths(1);
    end
end
% Determine PRS ID and coarse frequency offset by finding the strongest
% correlation
[fIdx,prsIdx] = find(peak_value==max(peak_value(:)));
coarseFrequencyOffset = fshifts(fIdx(1));

if do_plot
    % Plot PRS correlations and selected PRS ID
    figure;
    hold on;
    plot(fshifts/1e3,peak_value);
    title('PRS Correlations versus Frequency Offset');
    ylabel('Magnitude');
    xlabel('Frequency Offset (kHz)');
    plot(coarseFrequencyOffset/1e3,peak_value(fIdx(1),prsIdx(1)),'kx','LineWidth',2,'MarkerSize',8);
%     legends = "PRS ID" + num2cell(prsIDlist);
%     legend([legends "coarse $\Delta_f$ = " + num2str(coarseFrequencyOffset) + ", PRS ID = " + num2str(prsID)],'Location','SouthEast');
%     lgd = legend;
%     lgd.Interpreter = 'latex';
end
% Apply the coarse CFO correction
rxWaveformFreqCorrected = rxWaveform .* exp(-1i*2*pi*coarseFrequencyOffset*t);

freqOffset = coarseFrequencyOffset;
varargout{1} = freqOffset;
if doPRSSearch
    prsID = prsIDlist(prsIdx);
    varargout{2} = prsID;
end
end

File: PRS_CFO_estimate.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import correlate

def PRS_CFO_estimate(rxWaveform, refWaveform, TX_PRS_config, searchBW, prsIDlist, do_plot=False):
    """
    Performs a search for central frequency offset (CFO) and PRS ID (optionally).
    
    INPUTS:
        rxWaveform - received waveform, vector N_RXsym x 1
        refWaveform - reference waveform, vector N_RefSym x 1 or matrix N_RefSym x N_PRSid
        TX_PRS_config - transmit/reference signal configuration file
        searchBW - bandwidth for the frequency offset search
        prsIDlist - list for potential PRS IDs (if search is performed), or known PRS ID (if search is not performed)
        do_plot - indicator of plotting capability (True or False, False being default)
        
    OUTPUTS:
        rxWaveformFreqCorrected - frequency offset-corrected received waveform
        freqOffset - estimated frequency offset
        
    Optional output:
        prsID - found PRS ID (if search is performed)
    """
    
    if refWaveform.shape[1] == 1:
        doPRSSearch = False
    else:
        doPRSSearch = True

    # Set the CFO searching step
    fshifts = np.arange(-searchBW, searchBW + TX_PRS_config['subcarrierSpacing'][0][0], TX_PRS_config['subcarrierSpacing'][0][0]) * 1e3 / 2
    t = np.arange(rxWaveform.size) / TX_PRS_config['SampleRate'][0][0]
    
    peak_value = np.zeros((fshifts.size, refWaveform.shape[1])) if doPRSSearch else np.zeros(fshifts.size)
    peak_index = np.zeros((fshifts.size, refWaveform.shape[1])) if doPRSSearch else np.zeros(fshifts.size)

    for fIdx in range(fshifts.size):
        coarseFrequencyOffset = fshifts[fIdx]
        rxWaveformFreqCorrected = rxWaveform * np.exp(-1j * 2 * np.pi * coarseFrequencyOffset * t)
        T = rxWaveform.size
        
        if doPRSSearch:
            for prsIdx in range(refWaveform.shape[1]):
                refW = refWaveform[:, prsIdx]
                minlength = refWaveform.shape[0]
                
                if T < minlength:
                    waveformPad = np.concatenate([rxWaveformFreqCorrected, np.zeros(minlength - T)])
                    T = minlength
                else:
                    waveformPad = rxWaveformFreqCorrected
                
                refcorr = correlate(waveformPad, refW)
                corr = np.abs(refcorr[T-1:])
                peak_value[fIdx, prsIdx], peak_index[fIdx, prsIdx] = corr.max(), corr.argmax() + TX_PRS_config['SymbolLengths'][0][0]
        else:
            minlength = refWaveform.size
            
            if T < minlength:
                waveformPad = np.concatenate([rxWaveformFreqCorrected, np.zeros(minlength - T)])
                T = minlength
            else:
                waveformPad = rxWaveformFreqCorrected
                
            refcorr = correlate(waveformPad, refWaveform)
            corr = np.abs(refcorr[T-1:])
            peak_value[fIdx], peak_index[fIdx] = corr.max(), corr.argmax() + TX_PRS_config['SymbolLengths'][0][0]

    fIdx, prsIdx = np.unravel_index(np.argmax(peak_value, axis=None), peak_value.shape)
    coarseFrequencyOffset = fshifts[fIdx]

    if do_plot:
        plt.figure()
        plt.plot(fshifts / 1e3, peak_value)
        plt.title('PRS Correlations versus Frequency Offset')
        plt.ylabel('Magnitude')
        plt.xlabel('Frequency Offset (kHz)')
        plt.plot(coarseFrequencyOffset / 1e3, peak_value[fIdx, prsIdx] if doPRSSearch else peak_value[fIdx], 'kx', linewidth=2, markersize=8)
        plt.show()

    rxWaveformFreqCorrected = rxWaveform * np.exp(-1j * 2 * np.pi * coarseFrequencyOffset * t)
    freqOffset = coarseFrequencyOffset

    if doPRSSearch:
        prsID = prsIDlist[prsIdx]
        return rxWaveformFreqCorrected, freqOffset, prsID
    else:
        return rxWaveformFreqCorrected, freqOffset
